hazelcast/src/test/java/com/hazelcast/internal/nearcache/AbstractBasicNearCacheTest.java

```diff
@@ -0,0 +1,420 @@
+            executorService.execute(task);
+        }
+        assertOpenEventually(countDownLatch);
+
+        // the Near Cache is filled, we should see some memory costs now
+        if (context.hasLocalData && nearCacheConfig.getInMemoryFormat() != OBJECT) {
+            // the heap costs are just calculated if there is local data which is not in OBJECT in-memory-format
+            assertTrue("The Near Cache is filled, there should be some owned entry memory costs",
+                    context.stats.getOwnedEntryMemoryCost() > 0);
+            if (context.nearCacheAdapter.getLocalMapStats() != null) {
+                assertTrue("The Near Cache is filled, there should be some heap costs",
+                        context.nearCacheAdapter.getLocalMapStats().getHeapCost() > 0);
+            }
+        }
+
+        for (int i = 0; i < DEFAULT_RECORD_COUNT; i++) {
+            context.nearCacheAdapter.remove(i);
+        }
+        NearCacheTestUtils.waitForNearCacheSize(context, 0);
+
+        // the Near Cache is empty, we shouldn't see memory costs anymore
+        assertEquals("The Near Cache is empty, there should be no owned entry memory costs",
+                0, context.stats.getOwnedEntryMemoryCost());
+        if (context.nearCacheAdapter.getLocalMapStats() != null) {
+            // this assert will work in all scenarios, since the default value should be 0 if no costs are calculated
+            assertEquals("The Near Cache is empty, there should be no heap costs", 0,
+                    context.nearCacheAdapter.getLocalMapStats().getHeapCost());
+        }
+    }
+
+    /**
+     * Checks that the Near Cache eviction works as expected if the Near Cache is full.
+     */
+    @Test
+    public void testNearCacheEviction() {
+        NearCacheTestUtils.setEvictionConfig(nearCacheConfig, LRU, ENTRY_COUNT, DEFAULT_RECORD_COUNT);
+        NearCacheTestContext<Integer, String, NK, NV> context = createContext();
+
+        // all Near Cache implementations use the same eviction algorithm, which evicts a single entry
+        int expectedEvictions = 1;
+
+        // populate map with an extra entry
+        populateMap(context);
+        context.dataAdapter.put(DEFAULT_RECORD_COUNT, "value-" + DEFAULT_RECORD_COUNT);
+
+        // populate Near Caches
+        populateNearCache(context);
+
+        // we expect (size + the extra entry - the expectedEvictions) entries in the Near Cache
+        long expectedOwnedEntryCount = DEFAULT_RECORD_COUNT + 1 - expectedEvictions;
+        long expectedHits = context.stats.getHits();
+        long expectedMisses = context.stats.getMisses() + 1;
+
+        // trigger eviction via fetching the extra entry
+        context.nearCacheAdapter.get(DEFAULT_RECORD_COUNT);
+
+        NearCacheTestUtils.waitForNearCacheEvictions(context, expectedEvictions);
+
+        NearCacheTestUtils.assertNearCacheStats(context, expectedOwnedEntryCount, expectedHits, expectedMisses, expectedEvictions, 0);
+    }
```

Comment:
I'm slightly confused when a test has more than 3 blocks (like: init, action, assertion), It's tough to read its logic otherwise, but it's part of the personal style...